Lua support - initial PR

danpetry opened the pull request:
### Summary

This is an initial PR intended as a first step in development for adding Lua support. Lua is a lightweight scripting language suitable for embedded devices.

### Requirements

The requirements of this PR are to achieve the download and build into RIOT of the Lua interpreter, and demonstrate its working status via a simple example.

### Contribution details

#### Package

Lua v5.3.4 is cloned from github and built into RIOT, ready for its API to be accessed. Currently, only the Lua print() function is supported, which is adequate to demonstrate a working state of the interpreter.

#### Example

A .lua file called main.lua containing a simple "hello world" program, residing in the RIOT application's home directory, is converted to a char array at compile time, and stored in a header file called main.lua.h. This header file is then #included into an application, and the converted script is run in the program via a helper function called lua_run_script.

## Issues/PRs references

https://github.com/RIOT-OS/RIOT/issues/8434

danpetry commented:
@kaspar030 would you be interested in having a look at this?

jia200x commented on examples/lua/lua.c:
```diff
@@ -0,0 +1,658 @@
+
+#define LUA_INITVARVERSION	LUA_INIT_VAR LUA_VERSUFFIX
+
+
```
Please remove extra empty line here

jia200x commented on examples/lua/lua.c:
```diff
@@ -0,0 +1,658 @@
+#include "lauxlib.h"
+#include "lualib.h"
+
+#if !defined(LUA_PROMPT)
```
#ifndef?

jia200x commented on examples/lua/lua.c:
```diff
@@ -0,0 +1,658 @@
+/*
```
As far as I see, a large portion of this file is internal code. This should be in pkg/lua/contrib

danpetry replied:
A lot of this is implementation of a cli program, rather than the api - similar to e.g. examples/gcoap/gcoap_cli.c. So, is it best to think now about the API which is going to be exposed to an application and rearrange/separate out the lua.c file accordingly? My approach was to leave it to "evolve", but maybe the more usual approach is that code merged to master needs to be well thought out everywhere, is this true? If so will spend some time here...

danpetry replied:
Further thoughts: just spoke to @miri64 who suggested I think more about what the user would need to know about to implement a program (goes in example) and what they don't need to know (goes in pkg)

jcarrano replied:
I'm working on Lua now. The lua.c file is going away. It contains a REPL that is best implemented in Lua itself.

jia200x commented on examples/lua/main.c:
```diff
@@ -0,0 +1,53 @@
+}
+
+static const shell_command_t shell_commands[] = {
+    { "lua", "Start a Lua prompt", lua_main},
```
I think it would be good to have a single command here.
E.g `lua console`, `lua run`, etc.

jia200x commented on examples/lua/README.md:
```diff
@@ -0,0 +1,20 @@
+
+### How to run
+
+Type `make flash term
```
missing ` after "term"

jia200x commented on examples/lua/lua.c:
```diff
@@ -0,0 +1,658 @@
+
+#endif				/* } */
+
+
```
extra empty line

jia200x commented on examples/lua/lua.c:
```diff
@@ -0,0 +1,658 @@
+
+#endif				/* } */
+
+#endif				/* } */
```
Could you add /* */ to mark the closing bracket? E.g /* lua_stdin_is_tty */

jia200x commented on examples/lua/lua.c:
```diff
@@ -0,0 +1,658 @@
+
+#endif				/* } */
+
+
```
extra empty lines here

jia200x commented on examples/lua/Makefile:
```diff
@@ -0,0 +1,53 @@
+# example directory. The header file contains a byte array of the
+# ASCII characters in the .lua script.
+
+LUA_PATH := $(BINDIR)/lua
```
This lines can be moved to the pkg Makefiles and invoked when USEPKG+=lua. See e.g [here](https://github.com/RIOT-OS/RIOT/blob/master/pkg/openthread/Makefile#L16)

jia200x replied:
on second thoughts, maybe it's not bad to leave it as is. It explicitly tells the developer where's the file. But it could also be documented in the README.md. I'm OK with both ;)

danpetry replied:
I guess this wasn't too clear - this path variable is only used in this example, to contain the path where the script should be parsed to. I've renamed this LUA_SCRIPT_HEADER_PATH to avoid confusion and documented it in the readme. Also, it now builds to examples/lua/scripts. What do you think?

danpetry replied:
I'd prefer to keep it as it is, for consistency with examples/javascript and pkg/jerryscript which do the same thing.

jia200x replied:
ACK

jia200x replied:
maybe we can think about this later

danpetry replied:
Where's the best place for this kind of discussion after a PR has been closed?

jia200x replied:
Another PR/RFC should do the trick

jia200x commented on examples/lua/main.c:
```diff
@@ -0,0 +1,52 @@
+}
+
+static const shell_command_t shell_commands[] = {
+    { "run_lua_script", "Run main.lua", lua_run_main_script},
```
this command can be dropped, since there's no configuration required to run the main.lua script

jia200x commented on pkg/lua/Makefile:
```diff
@@ -0,0 +1,13 @@
+.PHONY: all
+
+all:
+#	The lower case "makefile" is so that it copies over the lua makefile
```
It's possible to specify a file with -f. It should be enough to copy the file

jia200x commented on pkg/lua/contrib/lua_run_char_array.c:
```diff
@@ -0,0 +1,155 @@
+static lua_State *globalL = NULL;
+
+static const char *progname = "lua";
+
```
Here and below, please remove double empty lines. Should be just one empty line between 2 blocks of code

jia200x commented on pkg/lua/contrib/lua_run_char_array.c:
```diff
@@ -0,0 +1,155 @@
+/*
+** Hook set by signal function to stop the interpreter.
+*/
+static void lstop (lua_State *L, lua_Debug *ar) {
```
although static, do you think it's possible to prefix these functions?

jia200x commented on pkg/lua/contrib/lua_run_char_array.c:
```diff
@@ -0,0 +1,155 @@
+** interpreter.
+*/
+static void laction (int i) {
+  signal(i, SIG_DFL); /* if another SIGINT happens, terminate process */
```
Identation should be 4 spaces

jia200x commented on examples/lua/main.c:
```diff
@@ -0,0 +1,52 @@
+#include "main.lua.h"
+
+extern int lua_main(int argc, char **argv);
+extern int lua_run_char_array(const char *buffer, size_t buffer_len);
```
This should be in a proper header file with it's Doxygen documentation. See e.g (here)[https://github.com/RIOT-OS/RIOT/blob/master/sys/include/base64.h#L56]

jia200x commented on examples/lua/main.c:
```diff
@@ -0,0 +1,52 @@
+
+#include <stdio.h>
+
+#include "msg.h"
```
why is msg.h required here?

jia200x commented:
@danpetry just wondering, I've check most of the functions are already implemented in external sources ([here](https://www.lua.org/source/5.3/lua.c.html) for instances). Is this file available in the Lua implementation? If so, maybe it could be a good idea to reuse as much as possible.

jia200x commented:
@danpetry what differences exists between their CLI and the one proposed here? If it's not part of the official CLI implementation, I think using an external file might help maintainance (if possible though).

What do you think?

jia200x commented:
As discussed offline with @danpetry, the original lua.c can be reused. Only some small functions in contrib folder would be required.

danpetry commented:
There are a few options for implementation of the pkg/contrib code

1. Have a cut-down version of lua.c in pkg/contrib

- Replicates some upstream code
- Avoids a patch, allowing code to be edited more readily for future development on this pkg
- Currently, this is what's done. (see pkg/contrib/lua_run_char_array.c)

2. patch the lua.c file.

- Some functions need to have their scope expanded by removing the static keyword. These are:

lstop - stops the interpreter
laction - sets a hook for lstop to stop the interpreter on a c signal
l_message - prints an error message
report - checks status and calls l_message if not ok
msghandler, docall, dochunk - run a chunk of lua code

- The lua_run_char_array function is then implemented in the pkg/contrib directory, and the functions above declared in that file as externs


Basically, it seems to be a choice between replication of upstream code or a patch. Opinions please? :)

danpetry commented:
So, here is another option which I think is better

The API wrapper code is not patched from Lua code, and it's not downstreamed either. It's re-written in a well-structured and maintainable way that can be extended in future iterations and included in the RIOT repo. It would need to be significantly different from the Lua code for this so neither patched nor downstreamed (nor a combination of both)

jia200x commented:
@danpetry can you provide a concrete example of this kind of wrapper?

If it's possible to call Lua instructions with the raw package, the current PR can be merged. I think the wrapper can come later, specially since there are other high level languages that might require also an abstraction layer.

danpetry commented:
ok so I'll move the implementation of the script parser to the example code?

jia200x commented:
@danpetry check the [current JS implementation](https://github.com/RIOT-OS/RIOT/tree/master/pkg/jerryscript). It fetches the original jerryscript code, and the application calls pkg functions.

I tend to think the Lua CLI can be added in another PR, so this one can easily gets merged with a simple example file showing actual Lua code. What do you think?

danpetry commented:
That sounds fine to me. Will do :)

jia200x commented on examples/lua/main.c:
```diff
@@ -0,0 +1,35 @@
+
+#include "main.lua.h"
+
+extern int lua_run_char_array(const char *buffer, size_t buffer_len);
```
This function should be included from a header file with its proper Doxygen documentation. As far as I see, it's intended to be called by the user

jia200x commented on examples/lua/main.c:
```diff
@@ -0,0 +1,58 @@
+
+    if (L == NULL) {
+        puts("cannot create state: not enough memory");
+        return EXIT_FAILURE;
```
@danpetry are this return codes LUA specific?

danpetry replied:
No, they're defined in stdlib.h

jia200x replied:
I would prefer here `errno` (e.g ENOBUFS)

danpetry replied:
I've used ENOMEM for this one; EINTR for failure of the protected call to the lua engine, to run the script; and `return 0;` for success in line with most other examples I looked at.

jia200x replied:
fine ;)

jia200x commented on examples/lua/main.c:
```diff
@@ -0,0 +1,58 @@
+
+#include "main.lua.h"
+
+int lua_run_script (const char *buffer, size_t buffer_len ){
```
please remove space between function name and parentheses, and between `buffer_len` and parentheses.
Also, add a space between the closing parentheses and curly bracket

danpetry replied:
Putting the opening curly bracket on the next line as I see that seems to be the most common way of doing it

jia200x replied:
yes, sorry. Curly bracket here should be in next line

jia200x commented on examples/lua/main.c:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    luaL_openlibs(L);
+
```
I would prefer to remove this extra line here

jia200x commented on examples/lua/main.c:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    lua_close(L);
+
```
also here

jia200x commented on examples/lua/main.c:
```diff
@@ -0,0 +1,58 @@
+int main(void)
+{
+    puts("Lua RIOT build");
+
```
and here

danpetry replied:
Thanks. Re the coding conventions, how much will uncrustify catch if I run that? Or is it more a case of learning by doing? (With some leeway re personal preference in certain cases?)

jia200x commented on pkg/lua/Makefile:
```diff
@@ -0,0 +1,13 @@
+PKG_NAME=lua
+PKG_URL=https://github.com/lua/lua.git
+PKG_VERSION=v5-3-4
```
This should pointing to a commit. See e.g [here](https://github.com/RIOT-OS/RIOT/blob/master/pkg/jerryscript/Makefile#L3)

jia200x commented:
@danpetry please address the comments above, squash and I give the ACK. I tested and works as expected.

Then I need @kaspar030 approval (or dismissing the review ;) in order to merge.

danpetry commented:
@kaspar030 said a few weeks ago he's ok with us proceeding, he won't have a look at it.

Addressing the rest of the comments now.

jia200x commented:
@danpetry Ok

danpetry commented:
I've made the changes, squashed, rebased, and pushed.
However, now it doesn't build for ARM boards, for the reasons detailed in https://github.com/RIOT-OS/RIOT/issues/9051. I narrowed the step which stopped it building down to the rebasing step. So, there must have been a change on RIOT master which now requires linking in some new functions. Going to do some more investigation, but if anyone has any helpful info or suggestions on how to proceed in the meantime, that would be helpful.
Initial thoughts are to get https://github.com/RIOT-OS/RIOT/pull/9090 merged, but we need to understand why it linked before and not now, in case there are other implications.

danpetry commented:
Update: it was this PR, which essentially removed some newlib stubs, which stops it building: https://github.com/RIOT-OS/RIOT/pull/8795

jia200x commented on pkg/lua/Makefile:
```diff
@@ -3,7 +3,7 @@ PKG_URL=https://github.com/lua/lua.git
 PKG_LICENSE=MIT
 PKG_BUILDDIR ?= $(BINDIRBASE)/pkg/$(BOARD)/$(PKG_NAME)
-
+#PKG_SOURCE_LOCAL=~/Sandbox/lua
```
Why is this line here?

jia200x commented on pkg/lua/Makefile.lua:
```diff
@@ -1,9 +1,13 @@
 
+# This redefines a lua prng function, avoiding the need for the _times system
+# call, which RIOT doesn't provide
+CFLAGS += -D'l_randomizePivot()=0'
```
I think it's a cleaner approach to remove this function with the patch

jcarrano replied:
The lua source is written in such a way that one is supposed to redefine things to change them. In this case, the definition of `l_randomizePivot()` is wrapped inside a `#ifndef l_randomizePivot`

danpetry replied:
I'm for putting all the patches for this fixup together in one patch. eg

`#define l_randomizePivot() 0`

just above `if !defined(l_randomizePivot)`

sorts it out.

jia200x commented on pkg/lua/patches/0001-Remove-the-Lua-functions-os.clock-and-os.rename.patch:
```diff
@@ -0,0 +1,39 @@
+-  const char *fromname = luaL_checkstring(L, 1);
+-  const char *toname = luaL_checkstring(L, 2);
+-  return luaL_fileresult(L, rename(fromname, toname) == 0, NULL);
++  return luaL_error(L, "This function is not implemented in RIOT yet");
```
This errors can be in a buffer, I don't think RIOT compiles with "string pooling".

jcarrano replied:
What's the advantage of putting the strings in buffers? String literals are all over the lua source code.

danpetry replied:
I agree with juan re this... I don't actually think it makes that much difference though.

jia200x replied:
in theory saves some ROM bytes, but if's the whole source code it won't make a big difference

jia200x replied:
ok, let's leave it like that then

jcarrano commented:
@jia200x @danpetry The next PR on lua will remove the REPL written in C (that is in lua.c) and replace it by a smaller and more elegant REPL written in Lua itself.

jia200x commented:
ok, works as expected. Tested in samr21-xpro and native. ACK

miri64 commented:
@aabadie @kaspar030 @smlng Murdock has an issue with this PR that we can't reproduce locally. Any idea what is wrong?

danpetry commented:
lauxlib.h is part of the pkg added as part of this PR. I suspect it's something to do with the way I've specified include paths. the two "debug commits" are there to get more info

miri64 commented:
> @miri64 I guess the murdock slave don't have the LUA-dev package? That's why:
> 
> ```
> fatal error: lauxlib.h: No such file or directory
> ```

@smlng I tested the same application with one of the failing boards on my dev machine and installed nothing extra (granted this is no guarantee that something isn't missing, but I most certainly don't have liblua-dev installed ;-)) and it worked.

```
$ apt list --installed | grep -i "lua"

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

liblua5.1-0/xenial,now 5.1.5-8ubuntu1 amd64 [installed,automatic]
liblua5.2-0/xenial,now 5.2.4-1ubuntu1 amd64 [installed,automatic]
libluajit-5.1-2/xenial,now 2.0.4+dfsg-1 amd64 [installed,automatic]
libluajit-5.1-common/xenial,xenial,now 2.0.4+dfsg-1 all [installed,automatic]
libtexlua52/xenial,now 2015.20160222.37495-1 amd64 [installed,automatic]
libtexluajit2/xenial,now 2015.20160222.37495-1 amd64 [installed,automatic]
lua-lpeg/xenial,now 0.12.2-1 amd64 [installed,automatic]
texlive-luatex/xenial-updates,xenial-updates,xenial-security,xenial-security,now 2015.20160320-1ubuntu0.1 all [installed,automatic]
```

aabadie commented on pkg/lua/Makefile.lua:
```diff
@@ -0,0 +1,9 @@
+# This builds for native using POSIX system calls and some extra libraries, and
+# removes a compiler warning that warns against using tmpnam().
+ifeq ($(BOARD),native)
+CFLAGS += -DLUA_USE_LINUX
```
minor, this should be indented with 2 spaces

miri64 replied:
A frustrated Dan just came to my office. Can we stop asking for minor changes when a PR is already ACK'd, please? This is also something that scares away contributors, IMHO. If the PR is defective and the ACK was given wrongly, yes then asking for changes is definitely okay. But not like this! Yes code quality is important, but these kind of non-functional errors can also be fixed in post.

aabadie replied:
I was pinged for a build issue, so I came, had a look at the package makefile, saw this and finally decided to **comment** this. I'm not requested any change.

tcschmidt replied:
Yep, and I guess @miri64 hit the point: Reviewers should carefully work with authors to get things into a state at which code can be merged. For this reviews need structuring, streamlining, and pacing in the most constructive manner. 
This means for instance that code-style review should be completed at some early stage, but also means to aid authors in case of troubles with build or similar.
I guess that's a generic discussion, but a very important one: I've seen many old open PRs where authors have been exhausted by being pushed around randomly without hitting anywhere.

danpetry replied:
I mean, I was just going to make the changes, as it would have been quicker than having a discussion about it. But the impression I get is that the PRs building up is a real problem and the review process itself should a) allow for valuable support of contributors by maintainers to get it into a good shape, which tbh has formed the bulk of my time on this PR; but also b) try not to duplicate work and streamline the process otherwise. I think it would be worth talking about tomorrow

miri64 replied:
> I was pinged for a build issue, so I came, had a look at the package makefile, saw this and finally decided to comment this. I'm not requesting any change.

Just to be clear, I appreciate that you only *commented* on this and did not make a change request.

> I didn't know an ACK by one reviewer should stop others from pointing out errors or possible improvements (as minor as they might be) anyway.

I already pointed out, that it's totally fine if the PR is defective, a late comment is ok. But it certainly gives a weird impression if a person that wasn't involved in the discussion before suddenly comes in and also comments on minor, non-functional details, when the PR was already ACK'd (even if they were pinged in for a completely different issue).

danpetry replied:
I think it's worth talking about this over voice call tomorrow... not sure this is a great format for this kind of discussion... :)

aabadie commented on examples/lua/Makefile:
```diff
@@ -0,0 +1,51 @@
+USEMODULE += ps
+
+ifneq ($(BOARD),native)
+# This stack size is large enough to run Lua print() functions of
```
indent this block (2 spaces)

jia200x commented:
ok, all green. Re-ACK.

smlng commented:
@jia200x then hit merge 🎉

jia200x commented:
>@jia200x then hit merge :tada:
![pic](https://imgflip.com/s/meme/Two-Buttons.jpg)

jia200x commented:
GO!

jia200x commented:
congratz @danpetry :tada: :tada:

danpetry commented:
Woop!
